test_dedupe.py

```
@@ -0,0 +1,95 @@
+        self.valves = valves
+
+    def normalize_text(self, t):
+        # Implementation from the modified adaptive_memory_v4.0.py
+        normalized = re.sub(r'[^\w\s]', '', t.strip().lower())
+        normalized = re.sub(r'\bs\b', '', normalized)
+        normalized = re.sub(r'\b(a|an|the)\b', '', normalized)
+        normalized = re.sub(r'\b(really|very|quite|pretty|so|totally|absolutely)\b', '', normalized)
+        normalized = re.sub(r'\s+', ' ', normalized).strip()
+        return normalized
+
+    async def _is_duplicate(self, text: str, memories: List[Any], exclude_id: str = None) -> bool:
```

Ruff (0.14.10)

PEP 484 prohibits implicit `Optional`

Convert to `T | None`

(RUF013)